### ISSUE TITLE
Why modin append method is very slow compared to pandas append?

### ISSUE THREAD

**Issue by Canon88** [modin-project/modin#620]
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: MacOS Mojave 10.14.4 (18E226)
- **Modin installed from (source or binary)**: pip install modin
- **Modin version**: 0.5.0
- **Python version**: 3.7
- **Exact command to reproduce**: 
```
for i in range(10):
    leaked = leaked.append(leaked)
```

### Describe the problem
When I used the append method for looping over data, the modin append method was considerably slower than the pandas append method. anyone tell me why? thanks!

e.g: 

Modin.append()
![image](https://user-images.githubusercontent.com/19372121/57744891-99c2e180-76fd-11e9-8deb-5686290509e9.png)

Pandas.append()
![image](https://user-images.githubusercontent.com/19372121/57745033-1786ed00-76fe-11e9-972d-e7c41f4a7878.png)

**Comment by devin-petersohn** [comment]
Thanks for the question @Canon88!

Essentially in Modin there is a cost to aligning the data when you have to shuffle between processes/partitions. Pandas does not have that cost because all the data is in the same process. Currently, operations like `reindex` are very expensive and are not 100% optimized. `append` uses the same logic as `reindex`, and we are planning to improve the performance of those operations after we have improved indexing performance.

This type of `append` in a loop is also an anti-pattern in both pandas and Modin. The best practices are to use `pd.concat` and pass the list of `DataFrame`s so that the metadata can be computed all at the same time, and pandas will also copy the data each time. There are fixed overheads to each `append` that are paid each time, and they are currently more expensive in Modin than pandas.

### Modin 36k rows x 4 cols
```
In [1]: import modin.pandas as pd 
   ...: import numpy as np 
   ...: frame_data = np.random.randint(0, 100, size=(36000, 4)) 
   ...: df = pd.DataFrame(frame_data).add_prefix("col")                                                                                           
2019-05-15 08:38:51,846	INFO node.py:497 -- Process STDOUT and STDERR is being redirected to /tmp/ray/session_2019-05-15_08-38-51_844278_58119/logs.
2019-05-15 08:38:51,955	INFO services.py:409 -- Waiting for redis server at 127.0.0.1:45965 to respond...
2019-05-15 08:38:52,067	INFO services.py:409 -- Waiting for redis server at 127.0.0.1:12859 to respond...
2019-05-15 08:38:52,069	INFO services.py:806 -- Starting Redis shard with 6.87 GB max memory.
2019-05-15 08:38:52,081	INFO node.py:511 -- Process STDOUT and STDERR is being redirected to /tmp/ray/session_2019-05-15_08-38-51_844278_58119/logs.
2019-05-15 08:38:52,082	INFO services.py:1441 -- Starting the Plasma object store with 20.0 GB memory using /tmp.
UserWarning: Distributing <class 'list'> object. This may take some time.

In [2]: df.shape                                                                                                                                  
Out[2]: (36000, 4)

In [3]: %%time 
   ...: for i in range(10): 
   ...:     df = df.append(df) 
   ...:                                                                                                                                           
CPU times: user 48 s, sys: 2.65 s, total: 50.6 s
Wall time: 1min 1s

In [4]: df = pd.DataFrame(frame_data).add_prefix("col")                                                                                           
UserWarning: Distributing <class 'list'> object. This may take some time.

In [5]: %%time 
   ...: df = pd.concat([df for _ in range(10)]) 
   ...:  
   ...:                                                                                                                                           
CPU times: user 211 ms, sys: 12.1 ms, total: 224 ms
Wall time: 226 ms

In [6]: df.shape                                                                                                                                  
Out[6]: (360000, 4)
```

### Pandas 36k rows x 4 cols
```
In [1]: import pandas as pd 
   ...: import numpy as np 
   ...: frame_data = np.random.randint(0, 100, size=(36000, 4)) 
   ...: df = pd.DataFrame(frame_data).add_prefix("col")                                                                                           

In [2]: %%time 
   ...: for i in range(10): 
   ...:     df = df.append(df) 
   ...:                                                                                                                                           
CPU times: user 888 ms, sys: 983 ms, total: 1.87 s
Wall time: 1.87 s

In [3]: df = pd.DataFrame(frame_data).add_prefix("col")                                                                                           

In [4]: %%time 
   ...: df = pd.concat([df for _ in range(10)]) 
   ...:  
   ...:                                                                                                                                           
CPU times: user 2.58 ms, sys: 1.79 ms, total: 4.37 ms
Wall time: 3.79 ms

In [5]: df.shape                                                                                                                                  
Out[5]: (360000, 4)
```
**For smaller DataFrames, these fixed overheads show more prevalently and currently Modin has higher overheads.**

## Modin 1m rows x 257 cols

```
In [1]: import modin.pandas as pd 
   ...: df = pd.read_csv("test_1mx256.csv")                                                                                                       
2019-05-15 08:46:43,294	INFO node.py:497 -- Process STDOUT and STDERR is being redirected to /tmp/ray/session_2019-05-15_08-46-43_293992_58150/logs.
2019-05-15 08:46:43,403	INFO services.py:409 -- Waiting for redis server at 127.0.0.1:14582 to respond...
2019-05-15 08:46:43,519	INFO services.py:409 -- Waiting for redis server at 127.0.0.1:22027 to respond...
2019-05-15 08:46:43,521	INFO services.py:806 -- Starting Redis shard with 6.87 GB max memory.
2019-05-15 08:46:43,531	INFO node.py:511 -- Process STDOUT and STDERR is being redirected to /tmp/ray/session_2019-05-15_08-46-43_293992_58150/logs.
2019-05-15 08:46:43,532	INFO services.py:1441 -- Starting the Plasma object store with 20.0 GB memory using /tmp.

In [2]: df.shape                                                                                                                                  
Out[2]: (1048576, 257)

In [3]: %%time 
   ...: df = pd.concat([df for _ in range(5)]) 
   ...:  
   ...:                                                                                                                                           
CPU times: user 233 ms, sys: 48.9 ms, total: 281 ms
Wall time: 2.6 s

In [4]: df.shape                                                                                                                                  
Out[4]: (5242880, 257)
```

## Pandas 1m rows x 257 cols

```
In [1]: import pandas as pd 
   ...: df = pd.read_csv("test_1mx256.csv")                                                                                                       

In [2]: %%time 
   ...: df = pd.concat([df for _ in range(5)]) 
   ...:  
   ...:                                                                                                                                           
CPU times: user 3.36 s, sys: 3.47 s, total: 6.83 s
Wall time: 6.84 s

In [3]: df.shape                                                                                                                                  
Out[3]: (5242880, 257)
```

TL;DR We know that shuffling operations are slow, and we are working on them, but `append` in a loop is always going to be worse in performance than `pd.concat`. At smaller datasets Modin will be slower, because there are fixed cost overheads to every operation.

**Comment by Canon88** [comment]
Thank you for your reply, I should use concat this method. Obviously, using modin when dealing with big data is even more advantageous. Thanks again!
